fix for issue #750; added check for token parameter

prahladyeri opened the pull request:
This is a fix for issue #750. What has changed:

- Added a check for `token` querystring parameter in `TkStart.java`.
- If token param is empty, proceeds to create a new bout as usual.
- If token param is not empty, searches it in the cache.
- If found in cache, gets the bout from there instead of creating new one.
- If not found in cache, creates a new bout and adds it to cache.

prahladyeri commented:
@dmzaytsev Please let me know whether my approach to fix this (PR #921 ) is correct or not since I know very less about `netbout`. Also, the build is mysteriously failing citing a `NullPointerException` on line 0 of TkStart (I haven't changed anything in header except adding a few imports):

    [ERROR] /src/main/java/com/netbout/rest/TkStart.java[0]: Got an exception - java.lang.NullPointerException (TreeWalker)

Can you help me with this?

dmzaytsev commented:
@prahladyeri It's Qulice issue not Netbout
>[INFO] --- qulice-maven-plugin:0.12:check (jcabi-check) @ netbout-web ---
[INFO] LICENSE found: file:C:/projects/netbout/netbout-web/LICENSE.txt 
[ERROR] \src\main\java\com\netbout\rest\TkStart.java[0]: Got an exception - java.lang.NullPointerException (TreeWalker) 
[INFO] Read our quality policy: http://www.qulice.com/quality.html 

feel free to submit a bug there 
https://github.com/teamed/qulice

prahladyeri commented:
Thanks @dmzaytsev . I've created a qulice [ticket](https://github.com/teamed/qulice/issues/608) for this.

karato commented:
@prahladyeri Thanks, let me find someone who can review this pull request

karato commented:
@carlosmiranda do this review

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +75,45 @@ public TkStart(final Base bse) {
+        if (tokenkey.equals("")) {
+            inbox = new RqAlias(this.base, req).alias().inbox();
+        } else {
+            if (this.tokencache.containsKey(tokenkey)) {
```
@prahladyeri I believe this `if-else` statement doesn't have to be nested, and can be moved to the same level as the outer `else`:

```java
if (tokenkey.equals("")) {
    tokenkey = token.next();
} else if (this.tokencache.containskey(tokenkey)) {
 // truncated code in inner if block
} else {
  // truncated code in inner else block
}
```

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +75,45 @@ public TkStart(final Base bse) {
+            inbox = new RqAlias(this.base, req).alias().inbox();
+        } else {
+            if (this.tokencache.containsKey(tokenkey)) {
+                final int hoursInDay = 24;
```
@prahladyeri let's use `TimeUnit` for these conversions. See [this](http://stackoverflow.com/a/19998563/4225089) for example.

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -58,6 +60,12 @@
     /**
+     * Token cache.
+     */
+    private final ConcurrentHashMap<String, Object[]> tokencache = new
```
@prahladyeri a couple of comments here:

1. Variable should be declared as `ConcurrentMap` interface type instead of using concrete class `ConcurrentHashMap`.
2. [Avoid compound names](http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html), we can call this `tokens` instead.

carlosmiranda commented:
@prahladyeri I have some comments related to the code. Please see above.

prahladyeri commented:
@carlosmiranda Thanks for your review. I've made all the changes except setting the ConcurrentMap value type to `Date[]`. You see, the value is an array of a `Date` and the bout inbox, that's why I've used `Object` data type.

The build is still failing because of a qulice issue which incorrectly flags a NullPointer exception on `line 0`, I've raised a separate [qulice bug](https://github.com/teamed/qulice/issues/608) for that which is in progress.

carlosmiranda commented:
@prahladyeri I don't think we should use `Object[]`, it's confusing and not type safe. Why not make the value a `Map<Date, Inbox>`?

carlosmiranda commented:
@prahladyeri Also we have the following checkstyle errors. They should be corrected or the merge will not be accepted:

```
[ERROR] /src/main/java/com/netbout/cached/CdInbox.java[86]: ')' have incorrect indentation level 4, expected level should be 8. (IndentationCheck)
[ERROR] /src/main/java/com/netbout/dynamo/DyInbox.java[179]: ')' have incorrect indentation level 4, expected level should be 8. (IndentationCheck)
[ERROR] /src/main/java/com/netbout/email/EmInbox.java[97]: ')' have incorrect indentation level 4, expected level should be 8. (IndentationCheck)
[ERROR] /src/main/java/com/netbout/rest/TkStart.java[0]: Got an exception - null (TreeWalker)
[ERROR] /src/test/java/com/netbout/rest/bout/XeMessageTest.java[73]: '.' have incorrect indentation level 9, expected level should be 12. (IndentationCheck)
```

prahladyeri commented:
@carlosmiranda There is only one checkstyle error which is causing the build to fail, rest were the result of upgrading the qulice-plugin to 0.13 (See [qulice issue 608](https://github.com/teamed/qulice/issues/608)). Now I've reverted that, and this is the only remaining exception which is incorrectly raised and qulice is working on that:

    [ERROR] \src\main\java\com\netbout\rest\TkStart.java[0]: Got an exception - java.lang.NullPointerException

Though `Object` is an unsafe type, I've wrapped it inside `ConcurrentHashMap` which makes it very much thread-safe. The problem in using `Map<Date, Inbox>` is that we cannot use `Date` as the key because we don't have the exact date while doing the comparison, we only difference it with the current date.

carlosmiranda commented:
@prahladyeri let's make `Inbox` the key instead. They're immutable and most implementations override `EqualsAndHashCode`. What do you think?

prahladyeri commented:
@carlosmiranda Again, `Inbox` is a thing we get from the `token cache` using the token key. The only pointer we have to fetch anything is the token key. Array is the only structure we can use here because the first element (`0`) can act as a key to fetch the `Date`, and second element (`1`) can act as the key to fetch the `Inbox`.

carlosmiranda commented:
@prahladyeri And we can do the same with a `Map`, except with type safety:

```
Inbox inbox = token.getKey();
Date date = token.getValue();
```

Or vice versa. Another alternative is we can make a simple `Token` object with the `Date` and `Inbox` objects. We are not restricted from `Object[]` which makes things confusing and unsafe.

dmzaytsev commented:
@prahladyeri let's exclude `TkStart` and create todo to remove the exclusion as @krzyk [suggested](https://github.com/teamed/qulice/issues/608#issuecomment-171210273)

> @dmzaytsev As per suggestion of krzyk, I tried with both 0.13 and 0.15.1 versions of maven qulice plugins, but neither of them works with this build.

prahladyeri commented:
@dmzaytsev Thanks for the suggestion. I've added `@todo` tasks accordingly.

As a side note, I've also clubbed the fix for #794 assigned to me in this PR (It is just a harmless text change in `README.md`). Please let me know whether its okay, or I have to revert everything and create a separate PR for that.

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +82,36 @@ public TkStart(final Base bse) {
+        final Iterator<String> token = href.param("token").iterator();
+        String tokenkey = "";
+        final Inbox inbox;
+        long number = 0;
```
@prahladyeri We can declare this in line 110 below, where it's assigned to the result of `inbox.start()`. That way we can make it final as well.

dmzaytsev replied:
@prahladyeri 
Exactly, you have to revert this here and create a new PR for the same
> As a side note, I've also clubbed the fix for #794 assigned to me in this PR (it is a harmless text change in README.md). Please let me know whether its fine, or I have to revert this and create a separate PR for that.

carlosmiranda commented:
@prahladyeri One additional comment above, plus you still have not addressed my comment regarding avoiding `Object[]` type. Please address them.

carlosmiranda commented:
@prahladyeri By the way, we don't mix tickets in a single PR, please create a separate branch for those changes.

prahladyeri commented:
@carlosmiranda Thanks, I've fixed this accordingly by moving the number declaration to line 110:

> We can declare this in line 110 below, where it's assigned to the result of inbox.start(). That way we can make it final as well.

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -49,6 +53,11 @@
+ * @todo #750:30min Solve the puzzle for build mysteriously failing for
+ *  this task citing a NullPointerException (See qulice issue #608 raised for
+ *  the same) and then remove TkStart.java from qulice exceptions.
+ * @todo #750:30min Solve the puzzle for needing an Object array in the
```
@prahladyeri I just had an idea - since we are importing `Commons-Lang3` we can use [Pair<Date, Inbox>](https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/tuple/Pair.html). Please update the code accordingly, I don't think we will need a puzzle for this.

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +82,35 @@ public TkStart(final Base bse) {
-        final Inbox inbox = new RqAlias(this.base, req).alias().inbox();
+        final Href href = new RqHref.Base(req).href();
+        final Iterator<String> token = href.param("token").iterator();
+        String tokenkey = "";
```
@prahladyeri Let's call this `key`, we should [avoid compound names](http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html).

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +82,35 @@ public TkStart(final Base bse) {
+        if (tokenkey.isEmpty()) {
+            inbox = new RqAlias(this.base, req).alias().inbox();
+        } else if (this.tokens.containsKey(tokenkey)) {
+            final Object[] cache = this.tokens.get(tokenkey);
```
@prahladyeri Let's call this `cached`, it better reflects what object it's referring to (the actual cache is `tokens`).

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +82,35 @@ public TkStart(final Base bse) {
+            inbox = new RqAlias(this.base, req).alias().inbox();
+        } else if (this.tokens.containsKey(tokenkey)) {
+            final Object[] cache = this.tokens.get(tokenkey);
+            final Date today = new Date();
```
@prahladyeri This is a single-use variable, we can inline it the line below.

carlosmiranda commented:
@prahladyeri Some more comments from me, then we should be good.

prahladyeri commented:
@carlosmiranda Done.

carlosmiranda commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -67,13 +81,35 @@ public TkStart(final Base bse) {
+        } else {
+            inbox = new RqAlias(this.base, req).alias().inbox();
+            this.tokens.put(key,
+                    Pair.of(new Date(), inbox));
```
@prahladyeri Let's move this param to the previous line, it shouldn't exceed length limit anymore.

carlosmiranda commented:
@prahladyeri One last comment related to formatting. :)

prahladyeri commented:
@carlosmiranda That too is done. :+1:

carlosmiranda commented:
@prahladyeri Excellent. I think you're picking up our process quickly. Your subsequent PR reviews should get shorter. :)

carlosmiranda commented:
@rultor merge please

rultor commented:
> @rultor merge please

@carlosmiranda Thanks for your request. @dmzaytsev Please confirm this.

dmzaytsev commented:
@prahladyeri as I see the token removed only if someone makes a request with same token two days later
but it may never happen
we will get memory leak in this case
I suggested to use guava cache because it allows to use an eviction policy
well you use the map as a token cache so you should remove tokens from the cache by the time not only on the next request. please create a puzzle for that. Tokens should be deleted in one hour after insert as Yegor requested.
there is no unit test for the new code as well. Now only tokenless requests tested.
We need a puzzle to test requests with tokens

prahladyeri commented:
@dmzaytsev For the time being, I've added the following little method which is called every hour and clears off all tokens older than one hour as per Yegor's recommendation:

    private void clearCache() {
        final Date now = new Date();
        long diff = 0;
        for (final String tkey : tokens.keySet()) {
            diff = now.getTime()
                - tokens.get(tkey).getLeft().getTime();
            if ((TimeUnit.MILLISECONDS.toHours(diff)) >= 1) {
                tokens.remove(tkey);
            }
        }
    }

However, I've added a `puzzle` to implement a proper Guava cache with an eviction policy sometime later. I've also added another `puzzle` for the unit tests:

     * @todo #750:30min Implement a proper Guava cache with eviction policy instead
     *  of this concurrent hashmap approach.
     * @todo #750:30min Add unit tests to test scenarios for requests
     *  with/without a token.

dmzaytsev commented on netbout-web/pom.xml:
```diff
@@ -733,6 +733,7 @@
                         <artifactId>qulice-maven-plugin</artifactId>
                         <configuration>
                             <excludes>
+                                <exclude>checkstyle:/src/main/java/com/netbout/rest/TkStart.java</exclude>
```
@prahladyeri did you try to remove the exclusion as @carlosmiranda [suggested](https://github.com/yegor256/netbout/pull/921#discussion_r49829400) ?

prahladyeri replied:
@dmzaytsev The suggestion by @carlosmiranda was for using `Pair` object from apache commons lang-3 (see again). Unfortunately, the exclusion isn't something we can presently remove (See [qulice issue 608](https://github.com/teamed/qulice/issues/608) raised for the same).

dmzaytsev commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -49,6 +54,13 @@
+ * @todo #750:30min Solve the puzzle for build mysteriously failing for
+ *  this task citing a NullPointerException (See qulice issue #608 raised for
+ *  the same) and then remove TkStart.java from qulice exceptions.
+ * @todo #750:30min Implement a proper Guava cache with eviction policy instead
```
@prahladyeri if you already implemented cache cleanup we don't need this puzzle 
lets's remove `TkStart#clearCache()` call from `TkStart#@act()`
and schedule it  like this
```
        public TkStart(final Base bse) {
            this.base = bse;
            Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(
                new Runnable() {
                    @Override
                    public void run() {
                        TkStart.this.clearCache();
                    }
                }, 1L, TimeUnit.MINUTES
            );
        }
```

dmzaytsev replied:
@prahladyeri let's remove this todo

dmzaytsev commented on netbout-web/src/main/java/com/netbout/rest/TkStart.java:
```diff
@@ -56,6 +68,17 @@
+    /**
+     * Last cache clearance time.
+     * */
+     private transient Date lastChecked = new Date();
```
@prahladyeri we don't need this variable if you will schedule the cleanup as above

dmzaytsev replied:
@prahladyeri let's get rid compound name

dmzaytsev commented:
@prahladyeri please see my comments

prahladyeri commented:
@dmzaytsev 
> lets's remove TkStart#clearCache() call from TkStart#@act()
and schedule it like this

Need a quick help regarding this. Firstly, you have given only three parameters for `scheduleAtFixedRate` method, but as per [official docs](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html) it has 4 even in Java 7. Are we using a very old version of java to build `netbout` ?

Secondly, my build is getting stuck at this point. Does this have anything to do with introducing the runnable?

        [INFO] --- jcabi-dynamodb-maven-plugin:0.7:create-tables (dynamodb-integration-test) @ netbout-web ---
        [INFO] >> 2016-01-16 15:28:08.339:INFO:oejs.Server:jetty-8.1.12.v20130726
        [INFO] >> 2016-01-16 15:28:08.447:INFO:oejs.AbstractConnector:Started SelectChannelConnector@0.0.0.0:51374
        [INFO] >> Jan 16, 2016 3:28:09 PM com.almworks.sqlite4java.Internal log
        [INFO] >> INFO: [sqlite] DB[1]: instantiated [AAAAABBBBBAAAAABBBBB_us-east-1.db]
        [INFO] >> Jan 16, 2016 3:28:09 PM com.almworks.sqlite4java.Internal log
        [INFO] >> INFO: [sqlite] Internal: loaded sqlite4java-linux-i386 from system path
        [INFO] >> Jan 16, 2016 3:28:09 PM com.almworks.sqlite4java.Internal log
        [INFO] >> INFO: [sqlite] Internal: loaded sqlite 3.7.10, wrapper 0.2
        [INFO] >> Jan 16, 2016 3:28:09 PM com.almworks.sqlite4java.Internal log
        [INFO] >> INFO: [sqlite] DB[1]: opened
        [INFO] Waiting for table 'nb-aliases' to become active
        [INFO] Table 'nb-aliases' is now ready for use
        [INFO] Waiting for table 'nb-friends' to become active
        [INFO] Table 'nb-friends' is now ready for use
        [INFO] Waiting for table 'nb-messages' to become active
        [INFO] Table 'nb-messages' is now ready for use
        [INFO] Waiting for table 'nb-attachments' to become active
        [INFO] Table 'nb-attachments' is now ready for use
        [INFO] 
        [INFO] --- exec-maven-plugin:1.3:java (start-takes-server) @ netbout-web ---
        [WARNING] Warning: killAfter is now deprecated. Do you need it ? Please comment on MEXEC-6.

dmzaytsev commented:
@prahladyeri I didn't check the code, perhaps I missed something, you can use parameters from jthe docs for sure

dmzaytsev commented:
@prahladyeri also please merge with master

dmzaytsev commented:
@prahladyeri Introducing the runnable shouldn't cause the stuck
if your `clearCache` works properly

prahladyeri commented:
> @prahladyeri Introducing the runnable shouldn't cause the stuck
if your clearCache works properly

No, that has to be something else. To test that hypothesis, I replaced my `clearCache()` method body with just a `return;` statement and it still gets stuck:

    private void clearCache() {
        return;
    }

prahladyeri commented:
@dmzaytsev We may not be able to use the `Runnable` I think. I'm getting the following PMD violation too on the runnable declaration line:

    TkStart.java[86-86]: To be compliant to J2EE, a webapp should not use any thread. (DoNotUseThreads)

Shall I create a todo for multi-threading and lets move on with the present code as it is?

dmzaytsev commented:
@prahladyeri let's keep it as is for now
please merge you branch with the latest master , remove the puzzle about Guava and rename `lastChecked`
and we are ready to merge

prahladyeri commented:
@dmzaytsev 
> please merge you branch with the latest master , remove the puzzle about Guava and rename lastChecked
and we are ready to merge

Done.

dmzaytsev commented:
@prahladyeri thank you!

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6369-172579003)

rultor commented:
> @rultor merge

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/6369-172579003) (took me 11min)

karato commented:
@ypshenychka please, review this ticket for compliance with our [QA rules](http://at.teamed.io/qa.html)

karato commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@karato OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6369-172975140)

rultor commented:
> @rultor please deploy

@karato Done! FYI, the full log is [here](http://www.rultor.com/t/6369-172975140) (took me 7min)

ypshenychka commented:
@karato Quality is good here.

karato commented:
> @karato Quality is good here.

@ypshenychka thanks, got it, great to see high-quality tickets here

karato commented:
@carlosmiranda paied 10 mins to @ypshenychka for QA review (payment ID is `75136501`). 10 mins added to @dmzaytsev account (our architect), payment ID is `75136510`, since quality is good here. I added **35 mins** to your account, many thanks for working with the project (75136518)! The completion time here was 151 hours and 58 mins.. review comments (c=20) added as a bonus. +35 added to your rating, current score is: [+2506](http://www.netbout.com/b/35023?open=rating)
